### PR TITLE
[state] validate system action layout

### PR DESCRIPTION
**Pull request by envestcc**
# Description
To avoid missing system actions in block generation, add validation of system actions when verifying received blocks.

## Type of change
Please delete options that are not relevant.
- [x] Bug fix (non-breaking change which fixes an issue)

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration
- [x] make test

**Test Configuration**:
- Firmware version:
- Hardware:
- Toolchain:
- SDK:

# Checklist:
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: state/factory/workingset.go [hunk 1]
```diff
@@ -404,6 +404,40 @@ func (ws *workingSet) process(ctx context.Context, actions []action.SealedEnvelo
 	return ws.finalize()
 }
 
+func (ws *workingSet) validateSystemActionLayout(ctx context.Context, actions []action.SealedEnvelope) error {
```
Comment (Liuhaai) [inline state/factory/workingset.go:407]: Reasons for moving it out of `process()`?

File: state/factory/workingset.go [hunk 2]
```diff
@@ -404,6 +404,40 @@ func (ws *workingSet) process(ctx context.Context, actions []action.SealedEnvelo
...
+	receivedSystemActions := []action.Envelope{}
+	startIdx := len(actions)
+	for i := range actions {
+		if action.IsSystemAction(actions[i]) {
+			if startIdx == len(actions) {
+				startIdx = i
+			}
+			receivedSystemActions = append(receivedSystemActions, actions[i].Envelope)
+		}
+	}
```
Comment (CoderZhi) [inline state/factory/workingset.go:428]: system actions should be at the end of the action list, and they should be continuous.

File: state/factory/workingset.go [hunk 3]
```diff
@@ -404,6 +404,40 @@ func (ws *workingSet) process(ctx context.Context, actions []action.SealedEnvelo
...
+	postSystemActions := []action.Envelope{}
+	for _, p := range reg.All() {
+		if psc, ok := p.(protocol.PostSystemActionsCreator); ok {
+			elps, err := psc.CreatePostSystemActions(ctx, ws)
+			if err != nil {
+				return err
+			}
+			postSystemActions = append(postSystemActions, elps...)
+		}
+	}
```
Comment (CoderZhi) [inline state/factory/workingset.go:418]: unify the generation of system actions (not signed) in a private function

**Review by CoderZhi — APPROVED** [review]

File: state/factory/workingset.go (hunk 1 above)
Comment (Liuhaai) [reply state/factory/workingset.go:407]: add comments for the func

File: state/factory/workingset.go [hunk 4]
```diff
@@ -404,6 +404,40 @@ func (ws *workingSet) process(ctx context.Context, actions []action.SealedEnvelo
 	return ws.finalize()
 }
 
+func (ws *workingSet) validateSystemActionLayout(ctx context.Context, actions []action.SealedEnvelope) error {
+	reg := protocol.MustGetRegistry(ctx)
+	postSystemActions := []action.Envelope{}
+	for _, p := range reg.All() {
+		if psc, ok := p.(protocol.PostSystemActionsCreator); ok {
+			elps, err := psc.CreatePostSystemActions(ctx, ws)
```
Comment (Liuhaai) [inline state/factory/workingset.go:412]: IMO, it's inefficient to regenerate the post-sys actions for each incoming block for reference. The actions should be generated and cached in the factory and be passed into `ValidateBlock()`.

**Review by Liuhaai — APPROVED** [review]

File: state/factory/workingset.go (hunk 1 above)
Comment (envestcc) [reply state/factory/workingset.go:407]: > add the purpose as the comment for the func

sure

File: state/factory/workingset.go (hunk 4 above)
Comment (envestcc) [reply state/factory/workingset.go:412]: > IMO, it's inefficient to regenerate the post-sys actions for each incoming block for reference. The actions should be generated and cached in the factory and be passed into `ValidateBlock()`.

The regeneration should be necessary for different heights. Is there a scenario where multiple blocks need to be verified at the same height?

File: state/factory/workingset.go (hunk 3 above)
Comment (envestcc) [reply state/factory/workingset.go:418]: > unify the generation of system actions (not signed) in a private function

OK

File: state/factory/workingset.go [hunk 5]
```diff
@@ -520,6 +554,9 @@ func (ws *workingSet) ValidateBlock(ctx context.Context, blk *block.Block) error
 	if err := ws.validateNonce(ctx, blk); err != nil {
 		return errors.Wrap(err, "failed to validate nonce")
 	}
+	if err := ws.validateSystemActionLayout(ctx, blk.RunnableActions().Actions()); err != nil {
```
Comment (dustinxie) [inline state/factory/workingset.go:568]: for safety, this should be gated by a hard-fork
then after chain passed the hard-fork height, we can remove the gating check

File: state/factory/workingset.go [hunk 6]
```diff
@@ -404,6 +404,47 @@ func (ws *workingSet) process(ctx context.Context, actions []action.SealedEnvelo
 	return ws.finalize()
 }
 
+func (ws *workingSet) generateSystemActions(ctx context.Context) ([]action.Envelope, error) {
```
Comment (CoderZhi) [inline state/factory/workingset.go:407]: https://github.com/iotexproject/iotex-core/blob/master/state/factory/factory.go#L362-L376
https://github.com/iotexproject/iotex-core/blob/master/state/factory/statedb.go#L242-L256

File: state/factory/workingset.go (hunk 4 above)
Comment (Liuhaai) [reply state/factory/workingset.go:412]: ok. it's a corner case. nervermind

File: state/factory/workingset.go [hunk 7]
```diff
@@ -404,6 +404,47 @@ func (ws *workingSet) process(ctx context.Context, actions []action.SealedEnvelo
...
+// validateSystemActionLayout verify whether the post system actions are appended tail
+func (ws *workingSet) validateSystemActionLayout(ctx context.Context, actions []action.SealedEnvelope) error {
+	postSystemActions, err := ws.generateSystemActions(ctx)
+	if err != nil {
+		return err
+	}
+	// system actions should be at the end of the action list, and they should be continuous
+	expectedStartIdx := len(actions) - len(postSystemActions)
+	offset := 0
+	for i := range actions {
```
Comment (dustinxie) [inline state/factory/workingset.go:431]:
```
if expectedStartIdx < 0 {
   return errors.Wrapf(errInvalidSystemActionLayout, ...
}
for i := expectedStartIdx; i < len(actions); i++ {
    if !action.IsSystemAction(actions[i]) {
        return errors.Wrapf(errInvalidSystemActionLayout, ...
    }
    if actions[i].Envelope.Proto().String() != {
        xxxx
    }
}
```
with this, only check `len(postSystemActions)` actions at the end, and no need to count `offset`

Comment (envestcc) [reply state/factory/workingset.go:431]: > ```
> if expectedStartIdx < 0 {
>    return errors.Wrapf(errInvalidSystemActionLayout, ...
> }
> for i := expectedStartIdx; i < len(actions); i++ {
>     if !action.IsSystemAction(actions[i]) {
>         return errors.Wrapf(errInvalidSystemActionLayout, ...
>     }
>     if actions[i].Envelope.Proto().String() != {
>         xxxx
>     }
> }
> ```
> 
> with this, only check `len(postSystemActions)` actions at the end, and no need to count `offset`

i think we also need to check whether the system action occurred prior to the expected start index

File: state/factory/workingset.go [hunk 8]
```diff
@@ -404,6 +404,47 @@ func (ws *workingSet) process(ctx context.Context, actions []action.SealedEnvelo
...
+		if action.IsSystemAction(actions[i]) {
+			if i != expectedStartIdx+offset {
+				return errors.Wrapf(errInvalidSystemActionLayout, "the action of idx %d should not be a system action", i)
```
Comment (dustinxie) [inline state/factory/workingset.go:434]: "the %d-th action"

File: state/factory/workingset.go [hunk 9]
```diff
@@ -404,6 +404,47 @@ func (ws *workingSet) process(ctx context.Context, actions []action.SealedEnvelo
...
+			}
+			if actions[i].Envelope.Proto().String() != postSystemActions[offset].Proto().String() {
+				return errors.Wrapf(errInvalidSystemActionLayout, "the action of idx %d is not the expected system action", i)
+			}
+			offset++
+		}
+	}
+	if offset != len(postSystemActions) {
+		return errors.Wrapf(errInvalidSystemActionLayout, "the number of system actions is incorrect, expected %d", len(postSystemActions))
```
Comment (dustinxie) [inline state/factory/workingset.go:443]: `expected %d, got %d", len(postSystemActions), offset)`

**Pull request merged**